File: hackit/xss/go/scanner.go
package main

import (
	"crypto/tls"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"
)

type Result struct {
	URL        string `json:"url"`
	Parameter  string `json:"parameter"`
	Payload    string `json:"payload"`
	Type       string `json:"type"`
	Details    string `json:"details"`
	Confidence string `json:"confidence"`
	Severity   string `json:"severity"`
	Impact     string `json:"impact"`
	Method     string `json:"method"`
}

type Scanner struct {
	Client   *http.Client
	Payloads []string
	Method   string
	Data     string
}

func NewScanner(timeout int) *Scanner {
	return &Scanner{
		Client: &http.Client{
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
			},
			Timeout: time.Duration(timeout) * time.Second,
		},
		Payloads: Payloads,
		Method:   "GET",
	}
}

func (s *Scanner) LoadPayloads(filePath string) error {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return err
	}
	lines := strings.Split(string(content), "\n")
	s.Payloads = make([]string, 0)
	for _, line := range lines {
		trimmed := strings.TrimSpace(line)
		if trimmed != "" {
			s.Payloads = append(s.Payloads, trimmed)
		}
	}
	return nil
}

func (s *Scanner) detectContext(bodyStr, pay string) (string, string, string, string) {
	bodyLower := strings.ToLower(bodyStr)
	payLower := strings.ToLower(pay)
	hasScriptOpen := strings.Contains(bodyLower, "<script>"+payLower) || strings.Contains(bodyLower, payLower+"</script>")
	hasAttrBreakout := strings.Contains(bodyLower, "=\""+payLower) || strings.Contains(bodyLower, "='"+payLower)
	hasHrefJS := strings.Contains(bodyLower, "href=\"javascript:") || strings.Contains(bodyLower, "src=javascript:")
	hasEventHandler := strings.Contains(bodyLower, "onerror="+payLower) || strings.Contains(bodyLower, "onload="+payLower) ||
		strings.Contains(bodyLower, "onclick="+payLower) || strings.Contains(bodyLower, "onfocus="+payLower) ||
		strings.Contains(bodyLower, "onmouseover="+payLower) || strings.Contains(bodyLower, "ontoggle="+payLower) ||
		strings.Contains(bodyLower, "onstart="+payLower)
	hasInScript := strings.Contains(bodyLower, "<script") && strings.Contains(bodyLower, payLower) && strings.Contains(bodyLower, "</script>")
	hasInEvent := strings.Contains(bodyLower, "on"+payLower)

	if hasHrefJS {
		return "CRITICAL: URI Handler Context (javascript:)", "High", "Critical", "Direct Code Execution"
	}
	if hasScriptOpen {
		return "CRITICAL: Executable Script Context", "High", "High", "Full Account Takeover (Session Theft)"
	}
	if hasEventHandler || hasInEvent {
		return "HIGH: Event Handler Context (executable)", "High", "High", "Sensitive Data Access / Redirection"
	}
	if hasInScript {
		return "HIGH: Inside Script Block", "High", "Medium", "Phishing / Content Injection"
	}
	if hasAttrBreakout {
		return "HIGH: Attribute Breakout Context", "High", "Medium", "Phishing / Forced Redirection"
	}
	if strings.Contains(pay, "{{") || strings.Contains(pay, "${") || strings.Contains(pay, "<%=") {
		return "MEDIUM: Template Injection Detected", "Medium", "Medium", "Server-Side Template Injection"
	}
	if strings.Contains(pay, "base64") || strings.Contains(pay, "data:") {
		return "MEDIUM: Data URI Injection", "Medium", "Medium", "Content Injection via Data URI"
	}
	if strings.Contains(pay, "constructor") || strings.Contains(pay, "__proto__") {
		return "MEDIUM: Prototype Pollution / Sandbox Escape", "Medium", "High", "Code Execution via Prototype Pollution"
	}
	return "LOW: Payload Reflected in Response", "Low", "Low", "Content Spoofing"
}

func (s *Scanner) isFalsePositive(bodyStr, pay string) bool {
	specificChars := []string{"<", ">", "\"", "'", "&"}
	for _, c := range specificChars {
		if strings.Contains(pay, c) {
			var encoded string
			switch c {
			case "<":
				encoded = "&lt;"
			case ">":
				encoded = "&gt;"
			case "\"":
				encoded = "&quot;"
			case "'":
				encoded = "&#39;"
			case "&":
				encoded = "&amp;"
			}
			if strings.Contains(bodyStr, encoded) && !strings.Contains(bodyStr, c) {
				return true
			}
		}
	}
	return false
}

func (s *Scanner) scanParam(urlStr string, param string, pay string, wg *sync.WaitGroup, sem chan struct{}, results *[]Result, mutex *sync.Mutex) {
	defer wg.Done()
	defer func() { <-sem }()

	u, err := url.Parse(urlStr)
	if err != nil {
		return
	}
	params := u.Query()
	newParams := make(url.Values)
	for k, v := range params {
		if k == param {
			newParams.Set(k, pay)
		} else {
			newParams[k] = v
		}
	}
	u.RawQuery = newParams.Encode()
	attackURL := u.String()

	req, err := http.NewRequest(s.Method, attackURL, nil)
	if err != nil {
		return
	}
	req.Header.Set("User-Agent", "HackIt-XSS/2.0")
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")

	resp, err := s.Client.Do(req)
	if err != nil {
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	bodyStr := string(body)

	if !strings.Contains(bodyStr, pay) {
		return
	}
	if s.isFalsePositive(bodyStr, pay) {
		return
	}

	details, confidence, severity, impact := s.detectContext(bodyStr, pay)

	if strings.Contains(pay, "cookie") || strings.Contains(pay, "fetch") || strings.Contains(pay, "xss.report") {
		severity = "Critical"
		impact = "Data Exfiltration (Sensitive Information)"
	}

	mutex.Lock()
	*results = append(*results, Result{
		URL:        attackURL,
		Parameter:  param,
		Payload:    pay,
		Type:       "Reflected XSS",
		Details:    details,
		Confidence: confidence,
		Severity:   severity,
		Impact:     impact,
		Method:     s.Method,
	})
	mutex.Unlock()
}

func (s *Scanner) Scan(targetURL string) []Result {
	results := make([]Result, 0)
	u, err := url.Parse(targetURL)
	if err != nil {
		return results
	}

	params := u.Query()
	if len(params) == 0 {
		fmt.Fprintf(os.Stderr, "[!] No query parameters in URL: %s\n", targetURL)
		return results
	}

	var wg sync.WaitGroup
	var mutex sync.Mutex
	sem := make(chan struct{}, 15)

	for param := range params {
		for _, payload := range s.Payloads {
			wg.Add(1)
			sem <- struct{}{}
			go s.scanParam(targetURL, param, payload, &wg, sem, &results, &mutex)
		}
	}
	wg.Wait()
	return results
}

File: hackit/xss/python/scanner.py
import json
import ssl
import urllib.request
import urllib.parse
import threading
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
from typing import Dict, Any, List, Optional

from .payloads import PAYLOADS


class Result:
    def __init__(self, url: str, parameter: str, payload: str,
                 details: str, confidence: str, severity: str,
                 impact: str, method: str = "GET"):
        self.url = url
        self.parameter = parameter
        self.payload = payload
        self.method = method
        self.type = "Reflected XSS"
        self.details = details
        self.confidence = confidence
        self.severity = severity
        self.impact = impact

    def to_dict(self) -> dict:
        return {
            "url": self.url,
            "parameter": self.parameter,
            "payload": self.payload,
            "method": self.method,
            "type": self.type,
            "details": self.details,
            "confidence": self.confidence,
            "severity": self.severity,
            "impact": self.impact,
        }


class Scanner:
    def __init__(self, timeout: int = 10, threads: int = 10):
        self.timeout = timeout
        self.threads = threads
        self.payloads = PAYLOADS
        self.ctx = ssl.create_default_context()
        self.ctx.check_hostname = False
        self.ctx.verify_mode = ssl.CERT_NONE

    def detect_context(self, body: str, pay: str) -> tuple:
        body_lower = body.lower()
        pay_lower = pay.lower()

        if "href=\"javascript:" in body_lower or "src=javascript:" in body_lower:
            return ("CRITICAL: URI Handler Context (javascript:)", "High", "Critical", "Direct Code Execution")
        if f"<script>{pay_lower}" in body_lower or f"{pay_lower}</script>" in body_lower:
            return ("CRITICAL: Executable Script Context", "High", "High", "Full Account Takeover (Session Theft)")
        event_keywords = ["onerror=", "onload=", "onclick=", "onfocus=", "onmouseover=", "ontoggle=", "onstart="]
        for kw in event_keywords:
            if kw + pay_lower in body_lower:
                return ("HIGH: Event Handler Context (executable)", "High", "High", "Sensitive Data Access / Redirection")
        if "<script" in body_lower and pay_lower in body_lower and "</script>" in body_lower:
            return ("HIGH: Inside Script Block", "High", "Medium", "Phishing / Content Injection")
        if f"=\"{pay_lower}" in body_lower or f"='{pay_lower}" in body_lower:
            return ("HIGH: Attribute Breakout Context", "High", "Medium", "Phishing / Forced Redirection")
        if "{{" in pay or "${" in pay or "<%=" in pay:
            return ("MEDIUM: Template Injection Detected", "Medium", "Medium", "Server-Side Template Injection")
        if "base64" in pay_lower or "data:" in pay_lower:
            return ("MEDIUM: Data URI Injection", "Medium", "Medium", "Content Injection via Data URI")

        return ("LOW: Payload Reflected in Response", "Low", "Low", "Content Spoofing")

    def is_false_positive(self, body: str, pay: str) -> bool:
        html_map = {"<": "&lt;", ">": "&gt;", "\"": "&quot;", "'": "&#39;", "&": "&amp;"}
        for char, encoded in html_map.items():
            if char in pay and encoded in body and char not in body:
                return True
        return False

    def test_payload(self, url: str, param: str, pay: str) -> Optional[Result]:
        try:
            parsed = urllib.parse.urlparse(url)
            params = urllib.parse.parse_qs(parsed.query, keep_blank_values=True)
            params[param] = [pay]
            new_query = urllib.parse.urlencode(params, doseq=True)
            attack_url = urllib.parse.urlunparse((
                parsed.scheme, parsed.netloc, parsed.path,
                parsed.params, new_query, parsed.fragment
            ))

            req = urllib.request.Request(
                attack_url,
                headers={"User-Agent": "HackIt-XSS/2.0", "Accept": "text/html,*/*"}
            )
            resp = urllib.request.urlopen(req, timeout=self.timeout, context=self.ctx)
            body = resp.read().decode('utf-8', errors='ignore')

            if pay not in body:
                return None
            if self.is_false_positive(body, pay):
                return None

            details, confidence, severity, impact = self.detect_context(body, pay)

            if "cookie" in pay.lower() or "fetch" in pay.lower():
                severity = "Critical"
                impact = "Data Exfiltration (Sensitive Information)"

            return Result(attack_url, param, pay, details, confidence, severity, impact)

        except Exception:
            return None

    def scan(self, target_url: str) -> List[Dict[str, Any]]:
        parsed = urllib.parse.urlparse(target_url)
        params = urllib.parse.parse_qs(parsed.query, keep_blank_values=True)
        if not params:
            return [{"error": "No query parameters in URL"}]

        results = []
        lock = threading.Lock()
        total = sum(len(self.payloads) for _ in params)
        done = 0

        def worker(param: str, pay: str):
            nonlocal done
            r = self.test_payload(target_url, param, pay)
            if r:
                with lock:
                    results.append(r.to_dict())
            with lock:
                done += 1

        with ThreadPoolExecutor(max_workers=self.threads) as executor:
            futures = []
            for param in params:
                for pay in self.payloads:
                    futures.append(executor.submit(worker, param, pay))
            for f in as_completed(futures):
                pass

        return results if results else [{"vulnerable": False, "note": "No XSS detected"}]
